File: solutions/erlang/pe_001_03.erl
% Project Euler <https://projecteuler.net>
% Problem: 1
% Solution: 3
% Author: Sean Broderick <hakutsuru@mac.com>
% Date: 2015-06-10


-module(pe_001_03).
-export([sum/1]).

% whole-number summation
% ((n^2)/2) + (n/2)
summation_to_number(X) ->
  ((X * X) + X) div 2.

sum(Threshold) ->
   Sum03 =  3 * summation_to_number(Threshold div 3),
   Sum05 =  5 * summation_to_number(Threshold div 5),
   Sum15 = 15 * summation_to_number(Threshold div 15),
   Sum03 + Sum05 - Sum15.


% Result (sum):
% /opt/euler/erlang$ erl
% 1> c(pe_001_03).
% {ok,pe_001_03}
% 2> pe_001_03:sum(9).
% 23
% 3> pe_001_03:sum(999).
% 233168
% 4> pe_001_03:sum(999999).
% 233333166668

File: solutions/elixir/pe_001_03.exs
# Project Euler <https://projecteuler.net>
# Problem: 1
# Solution: 3
# Author: Sean Broderick <hakutsuru@mac.com>
# Date: 2015-09-25


{threshold, _} = Integer.parse(List.first(System.argv))

# whole-number summation
# ((n^2)/2) + (n/2)
summation_to_number = fn x -> div(((x * x) + x), 2) end
   
sum03 =  3 * summation_to_number.(div(threshold, 3))
sum05 =  5 * summation_to_number.(div(threshold, 5))
sum15 = 15 * summation_to_number.(div(threshold, 15)) 

sum = sum03 + sum05 - sum15
IO.inspect sum


# Result (sum):
# /opt/euler/elixir$ elixir pe_001_03.exs 9
# 23
# /opt/euler/elixir$ elixir pe_001_03.exs 999
# 233168
# /opt/euler/elixir$ elixir pe_001_03.exs 999999
# 233333166668
